### PR TITLE
[TO MAIN] DESENG-513 - Add poll results to results tab

### PR DESCRIPTION
**Pull request by ratheesh-aot**
Issue #: https://apps.itsm.gov.bc.ca/jira/browse/DESENG-513

*Description of changes:*
  - Added poll results to the results tab.
  - Added poll results API.
  - Added Unit tests.

image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the met-public license (Apache 2.0).

### CODE REVIEW

File: met-web/src/services/widgetService/PollService/index.tsx [hunk 1]
```diff
@@ -73,3 +73,14 @@ export const postPollResponse = async (
         return Promise.reject(err);
     }
 };
+
+export const fetchPollResults = async (widget_id: number, poll_id: number): Promise<PollResultResponse> => {
+    try {
+        let url = replaceUrl(Endpoints.PollWidgets.RECORD_RESPONSE, 'widget_id', String(widget_id));
+        url = replaceUrl(url, 'poll_id', String(poll_id));
```
[the remark this answers is not in the document]
Comment (ratheesh-aot) [reply met-web/src/services/widgetService/PollService/index.tsx:80]: As replaceUrl only replaces one string at a time and here we need to replace both widget_id and poll_id, we need to call it twice.

File: met-web/src/services/widgetService/PollService/index.tsx [hunk 2]
```diff
@@ -73,3 +73,14 @@ export const postPollResponse = async (
...
+        const response = await http.GetRequest<PollResultResponse>(url);
+        return response.data || Promise.reject('Failed to fetch Poll Results');
+    } catch (err) {
+        return Promise.reject(err);
```
[the remark this answers is not in the document]
Comment (ratheesh-aot) [reply met-web/src/services/widgetService/PollService/index.tsx:84]: Agree. I followed the convention used in other service files.

File: met-web/src/services/widgetService/PollService/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Baelx) [reply met-web/src/services/widgetService/PollService/index.tsx:80]: Oh okay got it. Could `replaceAllInURL` be used to make it just one function call?

File: met-web/src/components/engagement/form/EngagementFormTabs/Results/PollResults/EngagementPollContext.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    const loadPollWidget = async () => {
+        const widget = widgets?.find((w) => w.widget_type_id === WidgetType.Poll) ?? null;
+        setWidget(widget);
+        if (!widget) {
+            setIsLoadingPollWidget(false);
+            return;
+        }
+        try {
+            const result = await fetchPollWidgets(widget.id);
+            setPollWidget(result[result.length - 1]);
```
[the remark this answers is not in the document]
Comment (ratheesh-aot) [reply met-web/src/components/engagement/form/EngagementFormTabs/Results/PollResults/EngagementPollContext.tsx:64]: Doable, i need to update the interface type also to accept undefined

File: met-web/src/components/engagement/form/EngagementFormTabs/Results/PollResults/EngagementPollContext.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        try {
+            const widgetsList = await getWidgets(savedEngagement.id);
+            setWidgets(widgetsList);
+            setIsWidgetsLoading(false);
+        } catch (err) {
+            setIsWidgetsLoading(false);
+            dispatch(openNotification({ severity: 'error', text: 'Error fetching engagement widgets' }));
+        }
```
[the remark this answers is not in the document]
Comment (ratheesh-aot) [reply met-web/src/components/engagement/form/EngagementFormTabs/Results/PollResults/EngagementPollContext.tsx:52]: Doable. Will update other useEffect functions too.

File: met-api/tests/unit/api/test_widget_poll.py [hunk 5]
```diff
@@ -211,3 +212,49 @@ def test_record_poll_response(client, session, jwt):
     )
 
     assert rv.status_code == HTTPStatus.BAD_REQUEST
+
+
+def test_get_poll_response(client, jwt, session, setup_admin_user_and_claims):
+    """Assert that a response for a poll widget can be retrived."""
```
[the remark this answers is not in the document]
Comment (ratheesh-aot) [reply met-api/tests/unit/api/test_widget_poll.py:218]: Thanks :)

**Pull request merged**